### PR TITLE
validation

### PR DESCRIPTION
**Pull request by LoganWright**
Validation proposal.

I added some updates that I think achieves some good syntax compromises between type safety, and basic accessors. I think we should discuss a lot of it, but here's the main points.

## Example

```Swift
class UserName: ValidationSuite {
    static func validate(input value: String) -> Bool {
        // composing validators. Can also be done inline
        let evaluation = OnlyAlphanumeric.self
            + StringLength.min(5)
            + StringLength.max(20)

        return evaluation.validate(value)
    }
}
```

Now Use

```Swift
let username = try possibleName.validated(by: UserName.self)
```

or w/ type inference

```Swift
class User {
    let username: Validated<UserName>

    init(username: String) throws {
        self.username = try username.validated()
    }
}
```

## Naming Conventions:

Some of the conventions I'm trying to stick to are as follows

### tested

Returns `Self` type, ie: `String` returns `String`, `Int` returns `Int`, etc. This is useful if one wants to incorporate validators, but not the `Validated<V>` type.

```Swift
// tested is String
let tested = try "someString".tested(by: ValidatorOne + ValidatorTwo || ValidatorThree)
```

### passes

Returns `Bool` whether or not a value is valid. I think I'd prefer to remove this, but I'm not sure if others think it has value.

### validated

This validates the input and creates a `Validated<V>` object that within reason can ensure the validity of the underlying value throughout its life.

```Swift
// Validated<Username>
let name = try "possibleName".validated(by: Username.self)
// Validated<And<Email, StringLength>>
let googleEmail = try "possible@email.com".validated(by: Email.from("google.com") + StringLength.min(13)) // Pretending google doesn't allow 1 or 2 letter emails (@google.com + 3)
```

## Composition

Validators can be easily composed together on the fly w/ logical operators `!`, `||`, and `+` or `&&`. 

```Swift
let secureEntry = try input.validated(by: StringLenght.min(5) + ContainsNumber.self || ContainsPunctuation.self)
```

#### Minor Annoyance

A lot of functions are based on `Validatable`. If a user wants a custom Validator, they need to make the `InputType` conform to `Validatable` with:

```Swift
extension CustomType: Validatable {}
```
It's simple to do, but a minor annoyance. This is what allows a flexible syntax w/o redeclaring the same functions for new types. Although it's annoying, I think it's better to the alternative. 

#### Value Safety

We had to sacrifice a bit of the value safety in order to have more dynamic validators. I think this is a good mix of the two concepts.

\- `ValidationSuite` which uses a class level initializer and can be more authoritative. This can't take parameters, but is a little bit stricter.

\- `Validator` initialized instance that takes parameters, but is slightly less safe. For example, `Validated<StringLength>` requires an instance of string length. We can guarantee the value has been validated by length, but we can't know the parameters passed. There is talk of eventually being able to do something like `StringLength<5>`. We should revisit this type when/if that becomes possible

## Discussion Points

- [x] Syntax. Do we like the way this looks
- [x] Naming Conventions, `tested`, `validated`, `passes`
- [x] Validatable Syntax. Possible alternatives
- [x] Operators. Should we include operators?
- [x] Should we keep them? Should use `+` or `&&` or both?

### CODE REVIEW

File: Sources/Vapor/Validation/Validation.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+@_exported import Validated
+
+// TODO: This is temporary to speed up dev, use real errors before :ship:
+extension String: ErrorProtocol {}
+
+extension Request.Content {
+    public func validated<ValidatorType: Validator>(key: String) throws -> Validated<ValidatorType> {
```
Comment (LoganWright) [inline Sources/Vapor/Validation/Validation.swift]: If we like the syntax of this function, we can work on figuring out the exact body in a way that works. Right now, it's essentially a place holder so we can test that things are compiling.

File: Package.swift [hunk 2]
```diff
@@ -17,7 +17,10 @@ let package = Package(
 
         //SHA2 + HMAC hashing. Used by the core to create session identifiers.
         .Package(url: "https://github.com/CryptoKitten/HMAC.git", majorVersion: 0, minor: 3),
-        .Package(url: "https://github.com/CryptoKitten/SHA2.git", majorVersion: 0, minor: 1)
+        .Package(url: "https://github.com/CryptoKitten/SHA2.git", majorVersion: 0, minor: 1),
+
+        //Validation type
+        .Package(url: "https://github.com/loganwright/validated.git", majorVersion: 3, minor: 0)
```
Comment (LoganWright) [inline Package.swift]: temporary until I can put up a PR to author. If author refuses changes, will fork to qutheory

**Comment by tannernelson** [comment]
That's a very interesting approach. A lot of cool functionality there.

To compare and contrast, here is another method.

```swift
public struct Example: RequestInitializable {
    let name: String
    let email: String
    let age: Int
    let bio: String
    let gender: Gender

    public init(_ request: Request) throws {
        name = try request.data["name"]?.string.verify(.min(5)) //name must be 5 chars long
        email = try request.data["email"]?.string.verify(.email) //must be email format
        age = try request.data["age"]?.int.verify(.min(21)) //must be at least 21 (different than string)
        bio = try request.data["bio"]?.string.verify(.min(50), .optional) //if provided, must be 50 chars long
        let gender = try request.data["gender"]?.string.verify(.in("m","f")) //no need to type hint gender as string
        self.gender = gender == "m" ? .male : .female
    }
}
```

Some new things here are being able to specify numbers for validation (min 50 characters, etc) as well as not having to type hint things (let gender is a string because `?.string`)

Cons
- `.data["bio"]?.string.verify` could be shorter.
- possible to get incorrect values (email is type String)

**Comment by LoganWright** [comment]
I added some updates that I think achieves some good syntax compromises between type safety, and basic accessors. I think we should discuss a lot of it, but here's the main points.

## Naming Conventions:

Some of the conventions I'm trying to stick to are as follows

### tested

Returns `Self` type, ie: `String` returns `String`, `Int` returns `Int`, etc. This is useful if one wants to incorporate validators, but not the `Validated<V>` type.

```Swift
let tested = try "someString".tested(by: ValidatorOne + ValidatorTwo || ValidatorThree)
```

### passes

Returns `Bool` whether or not a value is valid. This is useful if a user wants to combine multiple `Validator` types inside of an outer validator type:

```Swift
class UserName: ValidationSuite {
    static func validate(input value: String) -> Bool {
        let evaluation = OnlyAlphanumeric.self
            + StringLength.min(5)
            + StringLength.max(20)

        return value.passes(evaluation)
    }
}
```

### validated

This validates the input and creates a `Validated<V>` object that within reason can ensure the validity of the underlying value throughout its life.

```Swift
let name = try "possibleName".validated(by: Username.self)
let googleEmail = try "possible@email.com".validated(by: Email.from("google.com") + StringLength.min(3)) // Pretending google doesn't allow 1 or 2 letter emails
```

#### Minor Annoyance

A lot of functions are based on `Validatable`. If a user wants a custom Validator, they need to make the `InputType` conform to `Validatable` with:

```Swift
extension CustomType: Validatable {}
```
It's simple to do, but a minor annoyance. This is what allows a flexible syntax w/o redeclaring the same functions for new types. Although it's annoying, I think it's better to the alternative. 

## Discussion Points

- [ ] Syntax. Do we like the way this looks
- [ ] Naming Conventions, `tested`, `validated`, `passes`
- [ ] Validatable Syntax. Possible alternatives
- [ ] Operators. Should we include operators?
- [ ] Should we keep them? Should use `+` or `&&` or both?

File: Sources/Vapor/Validation/Validatable.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// MARK: Validation
+
+extension Validatable {
+    public func validated<V: Validator where V.InputType == Self>(by validator: V) throws -> Validated<V> {
+        return try Validated<V>(self, by: validator)
+    }
+    public func validated<S: ValidationSuite where S.InputType == Self>(by type: S.Type = S.self) throws -> Validated<S> {
+        return try Validated<S>(self, by: S.self)
```
Comment (LoganWright) [inline Sources/Vapor/Validation/Validatable.swift]: should take `type`. I don't see how they could be different, but I still think it's for the best

**Pull request merged**